### PR TITLE
Adding code for top level settings widget

### PR DESCRIPTION
**Pull request by skyler-rs**
> ### Motivate of the pull request
> -  Adds a top level settings widget that can find and display any settings dialogs that have been saved in settings
> image
> - Add the following to your task/setting dlg json to have it auto added to settings:
`            "_META_": {
                "isSetting": true
            }`
> - Logic has been added to alert the user when navigating away from a settings pane (changing categories in the main settings widget) to prompt them to save changes
> image

This currently doesn't have a menu pick in the UI, but can be invoked with 'EditSettings' in the console or 'EditSettings <task_name>' to auto select the given category.

### CODE REVIEW

File: src/Main/Settings.cpp [hunk 1]
```diff
@@ -103,6 +103,27 @@ void Settings::loadJsonFile(const QString& filePath) {
...
+// while storing the current path in path. The path can be used to create a
+// nlohmann::json_pointer which can be used to access the given node after this
+// traversal has finished
+void Settings::traverseJson(json& obj,
+                            std::function<void(json&, QString)> visitFn,
+                            QString path /* "" */) {
+  visitFn(obj, path);
+  if (obj.type() == nlohmann::detail::value_t::array) {
+    for (int i = 0; i < obj.size(); i++) {
+      QString childPath = path + "/" + QString::number(i);
```
Comment (KochynVolodymyr) [inline src/Main/Settings.cpp:116]: '/' might not work for WIN. Please check if it matters.

File: src/Main/Settings.cpp [hunk 2]
```diff
@@ -103,6 +103,27 @@ void Settings::loadJsonFile(const QString& filePath) {
...
+      traverseJson(obj.at(i), visitFn, childPath);
+    }
+  } else if (obj.type() == nlohmann::detail::value_t::object) {
+    for (auto& item : obj.items()) {
+      QString childPath = path + "/" + QString::fromStdString(item.key());
```
Comment (KochynVolodymyr) [inline src/Main/Settings.cpp:121]: '/' might not work for WIN. Please check if it matters.

File: src/Main/Settings.h [hunk 3]
```diff
@@ -53,6 +53,9 @@ class Settings {
       const QString& optionsArrayKey = "options",
       const QString& lookupArrayKey = "optionsLookup");
   QString getSystemDefaultSettingsDir();
+  static void traverseJson(json& obj,
+                           std::function<void(json&, QString)> visitFn,
+                           QString path = "");
```
Comment (KochynVolodymyr) [inline src/Main/Settings.h:58]: QString path = QString()

File: src/Main/WidgetFactory.cpp [hunk 4]
```diff
@@ -129,6 +137,199 @@ void setVal(QCheckBox* ptr, Qt::CheckState userVal) {
...
+            settingsWidget->layout()->setContentsMargins(0, 0, 0, 0);
+            settingsVLayout->addWidget(settingsWidget);
+
+            // Connect this dialog to the ok/cancel buttons of the child
+            // settings widget
+            if (settingsWidget) {
+              QDialogButtonBox* btnBox =
+                  settingsWidget->findChild<QDialogButtonBox*>(DlgBtnBoxName);
+              if (btnBox) {
+                QObject::connect(btnBox, &QDialogButtonBox::accepted, dlg,
```
[the remark this answers is not in the document]
Comment (skyler-rs) [reply src/Main/WidgetFactory.cpp:266]: This connecting to a dialog that gets created/destroyed each invocation so i believe the connections will all get cleaned up, but I have

File: src/Main/Settings.cpp (hunk 2 above)
Comment (skyler-rs) [reply src/Main/Settings.cpp:121]: This is a constructed path used by nlohmann::json, it won't interact with the file system so we don't have to worry about windows paths etc.

File: src/Main/Settings.h (hunk 3 above)
Comment (skyler-rs) [reply src/Main/Settings.h]: Fixed

**Comment by skyler-rs** [comment]
## Check-in Notes:
### Testing Status:
#### Current Testing:
The settings system is smoked tested with simple bring up tcl commands that just make sure the dialog comes up, doesn't crash, or leak data in valgrind.
#### Not Tested:
Currently the actual interaction with these dialogs as well as the persistence and tcl args interactions are not tested. New tests will be added once our framework provides the necessary controls.

### Json Library
The Settings system uses [nlohmann json](https://github.com/nlohmann/json), see the [docs](https://json.nlohmann.me/api/basic_json/#notes) for info on how to use the api.

**Comment by ravic-rs** [comment]
Functionality looks good, we will work on additional updating the tcl testing framework interface as a new feature.

**Pull request merged**